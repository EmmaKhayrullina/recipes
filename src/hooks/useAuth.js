import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { setUser, unsetUser } from '../store/actions/user';
import { register, logIn, signOut, initAuth } from '../services/authService';
import useAlert from './useAlert';
import { resetApp } from '../store/actions/app';

const useAuth = () => {
  const user = useSelector(state => state.user, shallowEqual);
  const dispatch = useDispatch();
  const { showAlertMessage } = useAlert();

  const signUpUser = async (email, password) => {
    try {
      const data = await register(email, password);
      const { email: userEmail, uid: userUid } = data.user;

      dispatch(setUser(userEmail, userUid));
      showAlertMessage({ text: `${userEmail} successfully registered!` });
    } catch (error) {
      showAlertMessage({ text: error.message, type: 'error' });
    }
  };

  const signInUser = async (email, password) => {
    try {
      const data = await logIn(email, password);
      const { email: userEmail, uid: userUid } = data.user;

      dispatch(setUser(userEmail, userUid));
      showAlertMessage({ text: `Hello ${userEmail}!` });
    } catch (error) {
      showAlertMessage({ text: error.message, type: 'error' });
    }
  };

  const checkUserSession = () => {
    return initAuth(userData => {
      if (userData && !user.uid) {
        dispatch(setUser(userData.email, userData.uid));
      }
    });
  };

  const signOutUser = async () => {
    await signOut().then(() => {
      dispatch(unsetUser());
      dispatch(resetApp());
      showAlertMessage({ text: `See you soon!` });
    });
  };

  return {
    user,
    signUpUser,
    signInUser,
    signOutUser,
    checkUserSession,
  };
};

export default useAuth;
