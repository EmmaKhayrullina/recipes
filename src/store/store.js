import { createStore } from 'redux';
import rootReducer from './reducers/rootReducer';

const saveToLocalStorage = state => {
  try {
    localStorage.setItem('userData', JSON.stringify(state));
  } catch (e) {
    throw new Error(e);
  }
};

const loadFromLocalStorage = () => {
  try {
    const userData = localStorage.getItem('userData');
    if (userData === null) return undefined;
    return JSON.parse(userData);
  } catch (e) {
    throw new Error(e);
  }
};

const persistedState = loadFromLocalStorage();

/* eslint-disable no-underscore-dangle */
const store = createStore(
  rootReducer,
  persistedState,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
);
/* eslint-enable no-underscore-dangle */

store.subscribe(() => saveToLocalStorage(store.getState()));

export default store;
