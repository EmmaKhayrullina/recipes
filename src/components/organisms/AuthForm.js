import React, { useState } from 'react';
import { Link, withRouter } from 'react-router-dom';
import styled from 'styled-components';
import Button from '../atoms/Button';
import Input from '../atoms/Input';
import Form from '../atoms/Form';
import useAuth from '../../hooks/useAuth';

const FieldActions = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
`;

const AuthForm = props => {
  const {
    user: { email },
    pageName,
    history,
  } = props;
  const [user, setUser] = useState({ email, password: '' });
  const { signUpUser, signInUser } = useAuth();

  const onChangeField = event => {
    const { name, value } = event.target;
    setUser(userData => ({ ...userData, [name]: value }));
  };

  const redirectToHome = () => history.replace('/');

  const onSubmitHandler = e => {
    if (e) e.preventDefault();

    if (pageName === 'Login') {
      return signInUser(user.email, user.password, redirectToHome);
    }
    return signUpUser(user.email, user.password, redirectToHome);
  };

  return (
    <Form onSubmit={onSubmitHandler} data-testid={`test-${pageName.toLowerCase()}-form`}>
      <Input type="email" label="Email" name="email" value={user.email} onChange={onChangeField} />
      <Input type="password" label="Password" name="password" value={user.password} onChange={onChangeField} />

      <FieldActions>
        {pageName === 'Login' ? <Link to="/registration">Registration</Link> : <Link to="/login">Login</Link>}

        <Button type="submit">{pageName}</Button>
      </FieldActions>
    </Form>
  );
};

export default withRouter(AuthForm);
