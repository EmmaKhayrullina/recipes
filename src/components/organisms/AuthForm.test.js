import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../../store/store';
import AuthForm from './AuthForm';
import useAuth from '../../hooks/useAuth';

jest.mock('../../hooks/useAuth');

const renderAuthForm = (pageName, user) => {
  return render(
    <Provider store={store}>
      <MemoryRouter>
        <AuthForm user={user} pageName={pageName} />
      </MemoryRouter>
    </Provider>,
  );
};

describe('AuthForm component', () => {
  const signUp = jest.fn();
  const signIn = jest.fn();
  const mockUseAuth = jest.fn(() => ({
    signUpUser: signUp,
    signInUser: signIn,
  }));

  useAuth.mockImplementation(mockUseAuth);

  beforeEach(() => {
    const pageName = 'Login';
    const user = { email: '', uid: '' };
    renderAuthForm(pageName, user);
  });

  test('should render login form', () => {
    // Assert
    expect(screen.getByTestId('test-login-form')).toMatchSnapshot();
  });

  test('should submit login form', () => {
    // Arrange
    const { signInUser } = useAuth();

    // Act
    fireEvent.submit(screen.getByRole('button'), { name: /login/i });

    // Assert
    expect(signInUser).toHaveBeenCalled();
  });

  test('should render registration form', () => {
    // Arrange
    const pageName = 'Registration';
    const user = { email: 'test@test.com', uid: '123456' };

    // Act
    renderAuthForm(pageName, user);

    // Assert
    expect(screen.getByTestId('test-registration-form')).toMatchSnapshot();
  });

  test('should register user', () => {
    // Arrange
    const { signUpUser } = useAuth();
    const pageName = 'Registration';
    const user = { email: 'test@test.com', uid: '123456' };

    renderAuthForm(pageName, user);

    // Act
    fireEvent.submit(screen.getByRole('button', { name: /registration/i }));

    // Assert
    expect(signUpUser).toHaveBeenCalled();
  });

  test('should change fields', () => {
    // Arrange
    const expectedEmailValue = 'test@test.com';
    const expectedPasswordValue = '123456';
    const email = screen.getByRole('textbox', { name: /email/i });
    const password = screen.getByLabelText(/password/i);

    // Act
    fireEvent.change(email, { target: { name: 'email', value: expectedEmailValue } });
    fireEvent.change(password, { target: { name: 'password', value: expectedPasswordValue } });

    // Assert
    expect(email.value).toEqual(expectedEmailValue);
    expect(password.value).toEqual(expectedPasswordValue);
  });
});
