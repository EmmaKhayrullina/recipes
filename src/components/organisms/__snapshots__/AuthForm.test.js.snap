// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm component should render login form 1`] = `
<form
  class="sc-AxhCb louZza"
  data-testid="test-login-form"
>
  <div
    class="sc-AxirZ ifDUjn"
  >
    <label
      class="sc-AxiKw bpwopD"
      for="input-email"
    >
      Email
    </label>
    <input
      id="input-email"
      name="email"
      type="email"
      value=""
    />
    <span />
  </div>
  <div
    class="sc-AxirZ ifDUjn"
  >
    <label
      class="sc-AxiKw bpwopD"
      for="input-password"
    >
      Password
    </label>
    <input
      id="input-password"
      name="password"
      type="password"
      value=""
    />
    <span />
  </div>
  <div
    class="sc-AxhUy hPJHzO"
  >
    <a
      href="/registration"
    >
      Registration
    </a>
    <button
      class="sc-AxjAm kEbvLn"
      type="submit"
    >
      Login
    </button>
  </div>
</form>
`;

exports[`AuthForm component should render registration form 1`] = `
<form
  class="sc-AxhCb louZza"
  data-testid="test-registration-form"
>
  <div
    class="sc-AxirZ ifDUjn"
  >
    <label
      class="sc-AxiKw bpwopD"
      for="input-email"
    >
      Email
    </label>
    <input
      id="input-email"
      name="email"
      type="email"
      value="test@test.com"
    />
    <span />
  </div>
  <div
    class="sc-AxirZ ifDUjn"
  >
    <label
      class="sc-AxiKw bpwopD"
      for="input-password"
    >
      Password
    </label>
    <input
      id="input-password"
      name="password"
      type="password"
      value=""
    />
    <span />
  </div>
  <div
    class="sc-AxhUy hPJHzO"
  >
    <a
      href="/login"
    >
      Login
    </a>
    <button
      class="sc-AxjAm kEbvLn"
      type="submit"
    >
      Registration
    </button>
  </div>
</form>
`;
